read-fonts/src/tables/colr.rs

```diff
@@ -1,5 +1,696 @@
+                    extend,
+                }
+            }
+            Self::RadialGradient(gradient) => {
+                let color_line = gradient.color_line()?;
+                let extend = color_line.extend();
+                ResolvedPaint::RadialGradient {
+                    x0: gradient.x0().to_fixed(),
+                    y0: gradient.y0().to_fixed(),
+                    radius0: gradient.radius0().to_fixed(),
+                    x1: gradient.x1().to_fixed(),
+                    y1: gradient.y1().to_fixed(),
+                    radius1: gradient.radius1().to_fixed(),
+                    color_stops: color_line.into(),
+                    extend,
+                }
+            }
+            Self::VarRadialGradient(gradient) => {
+                let color_line = gradient.color_line()?;
+                let extend = color_line.extend();
+                let deltas = instance.var_deltas::<6>(gradient.var_index_base());
+                ResolvedPaint::RadialGradient {
+                    x0: gradient.x0().apply_delta(deltas[0]),
+                    y0: gradient.y0().apply_delta(deltas[1]),
+                    radius0: gradient.radius0().apply_delta(deltas[2]),
+                    x1: gradient.x1().apply_delta(deltas[3]),
+                    y1: gradient.y1().apply_delta(deltas[4]),
+                    radius1: gradient.radius1().apply_delta(deltas[5]),
+                    color_stops: color_line.into(),
+                    extend,
+                }
+            }
+            Self::SweepGradient(gradient) => {
+                let color_line = gradient.color_line()?;
+                let extend = color_line.extend();
+                ResolvedPaint::SweepGradient {
+                    center_x: gradient.center_x().to_fixed(),
+                    center_y: gradient.center_y().to_fixed(),
+                    start_angle: gradient.start_angle().to_fixed(),
+                    end_angle: gradient.end_angle().to_fixed(),
+                    color_stops: color_line.into(),
+                    extend,
+                }
+            }
+            Self::VarSweepGradient(gradient) => {
+                let color_line = gradient.color_line()?;
+                let extend = color_line.extend();
+                let deltas = instance.var_deltas::<4>(gradient.var_index_base());
+                ResolvedPaint::SweepGradient {
+                    center_x: gradient.center_x().apply_delta(deltas[0]),
+                    center_y: gradient.center_y().apply_delta(deltas[1]),
+                    start_angle: gradient.start_angle().apply_delta(deltas[2]),
+                    end_angle: gradient.end_angle().apply_delta(deltas[3]),
+                    color_stops: color_line.into(),
+                    extend,
+                }
+            }
+            Self::Glyph(glyph) => ResolvedPaint::Glyph {
+                glyph_id: glyph.glyph_id(),
+                paint: glyph.paint()?,
```

Comment:
is it intentional that this does not resolve all sub-paints recursively? probably so, but I figured I'd ask.
Is it even possible for a Rust struct (e.g. ResolvedPaint) to have a field that is of the same type as itself?

Reply:
Yes, this is intentional. Fully resolving the graph up front would require boxing the recursive type.

The relationship between `Paint` and `ResolvedPaint` is similar to `FT_OpaquePaint` and `FT_COLR_Paint`.